cilium/CFP-30788-managing-hubble-metrics-cardinality.md

```diff
@@ -0,0 +1,135 @@
+
+## Goals
+
+* Ability to manage the cardinality of Hubble metrics dynamically.
+* Should be able to control labels and dimensions of the metrics.
+
+
+## Non-Goals
+
+* Recommendations for Prometheus configuration at ingestion time.
+
+## Proposal
+
+### Overview
+
+We propose a new config map wrapped functionality in Hubble metrics which can be used to manage the cardinality, dimensions and targets of the metrics. Drawing from [Hubble Flows dynamic exporter](https://docs.cilium.io/en/latest/observability/hubble-exporter/) which uses a standalone config map to manage exporting of flows, we propose a similar mechanism. 
+
+Existing behavior of reading cilium-config for `hubble-metrics` be preserved for backward compatibility. 
+
+A new config called `enable-dynamic-hubble-metrics` be introduced which when set to true, will enable the dynamic management of Hubble metrics. When this config is set to true, the cilium agent will read the `hubble-metrics-config` config map to get the configuration for the metrics. When this config is set to false, the cilium agent will read the `cilium-config` for `hubble-metrics` configuration.
+
+### Config Map
+
+A new config map called `hubble-metrics-config` will be introduced which will have the following fields:
+
+- `enabled`: Boolean value to enable/disable the dynamic management of Hubble metrics.
+- `metrics`: List of metrics to be managed. Each metric will have the following fields:
+  - `name`: Name of the metric.
+  - `contextOptions`: List of [context options](https://docs.cilium.io/en/latest/observability/metrics/#context-options) to be applied to the metric.
+  - `includeFilters` : List of include flow filters to be applied to this particular metric.
+  - `excludeFilters` : List of exclude flow filters to be applied to this particular metric.
+
+### Example
+    
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+    name: hubble-metrics-config
+    namespace: kube-system
+data:
+    hubblemetrics.yaml: |
+        hubbleMetrics:
+            enabled: true
+            metrics:
+            - name: flow
+                contextOptions: 
+                    - sourceEgressContext=pod
+                    - destinationIngressContext=pod
+                includeFilters: 
+                    - []
+                excludeFilters:
+                    - source_pod: ["namespace1/"]
+                    - destination_pod: ["kube-system/"]
+            - name: dns
+                contextOptions: 
+                    - labelsContext=source_namespace,source_pod
+                    - destinationIngressContext=pod
+                includeFilters: 
+                    - source_pod: ["namespace1/"]
+```
+
+This will make sure we will have a dynamic way to manage each metric's cardinality, dimensions and targets. The flow filters can  be applied on each metric's `ProcessFlow` function save CPU cyles.
```

Comment:
IMO implementation details should be discussed in the follow up PRs rather than at the design stage.

```suggestion
This will make sure we will have a dynamic way to manage each metric's cardinality, dimensions and targets.
```